### PR TITLE
Feature/forms

### PR DESCRIPTION
**Pull request by phfdonda**
**Hi!**
This is the conclusion of the Re-form project about Form in Rails. We have:
1- Created a new Rails app;
2- Generated a simple User controller, model and view;
3- Applied the new knowledge about "form_for";
4- Created an error message about wrong parameters in forms.
We hope this pull request is of your liking, let us know if we can improve it in any way.
Thank you for your review!

### CODE REVIEW

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Changes Required 🔍 

Hi Team! 👋 

Good job!

#### Required Changes
- You seem not to have your stickler set up, please do this. See this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/linter_config.png) from [Pathwright](https://microverse.pathwright.com/auth/sign-in/?next=/library/fast-track-curriculum/69047/path/step/54883773/)
- You are missing the commented out `form_tag` form as can be seen from this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_form_tag2.png) from Odin.

***Please makes changes and submit another review request.
If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  private
+
+  def user_params
+    params.require(:user).permit(:username, :email, :password)
+  end
+
+  def set_user
+    @user = User.find(params[:id])
+  end
+
```
[the remark this answers is not in the document]
Comment (phfdonda) [reply app/controllers/users_controller.rb]: Solved

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+
+class EmailValidator < ActiveModel::EachValidator
```
[the remark this answers is not in the document]
Comment (phfdonda) [reply app/models/user.rb:1]: Solved

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  def validate_each(record, attribute, value)
+    unless value =~ /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i
```
[the remark this answers is not in the document]
Comment (phfdonda) [reply app/models/user.rb]: Solved

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      record.errors[attribute] << (options[:message] || "is not an email")
```
[the remark this answers is not in the document]
Comment (phfdonda) [reply app/models/user.rb]: Solved

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    end
+  end
+end
+
+class AlphanumericValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    if value.nil?
```
[the remark this answers is not in the document]
Comment (phfdonda) [reply app/models/user.rb]: Solved

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      record.errors[attribute] << (options[:message] || "value must not be nil")
```
[the remark this answers is not in the document]
Comment (phfdonda) [reply app/models/user.rb]: Solved

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    end
+    unless value.split('').none?(/[\W]/)
```
[the remark this answers is not in the document]
Comment (phfdonda) [reply app/models/user.rb]: Solved

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+      record.errors[attribute] << (options[:message] || "must be composed of only letters and numbers")
```
[the remark this answers is not in the document]
Comment (phfdonda) [reply app/models/user.rb]: Solved

File: app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    end
+  end
+end
+
+class User < ApplicationRecord
+
```
[the remark this answers is not in the document]
Comment (phfdonda) [reply app/models/user.rb]: Solved

File: config/routes.rb [hunk 10]
```diff
@@ -1,3 +1,4 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  resources :users, only: [:new, :create, :edit, :update]
```
[the remark this answers is not in the document]
Comment (phfdonda) [reply config/routes.rb]: Solved

File: config/routes.rb [hunk 11]
```diff
@@ -1,3 +1,4 @@
...
+
```
[the remark this answers is not in the document]
Comment (phfdonda) [reply config/routes.rb]: Solved

**Comment by phfdonda** [comment]
Hi!
I solved all errors reported. I didn't get that I should keep the form_tag commented, thanks for pointing that out.
I hope everything's ok now, let me know if there's anything else I should fix.
Thank you for your review!

**Comment by uryelah** [comment]
Hi **Pedro Donda & J. P. Pérez Tejada Ladrón de Guevara**, I'm Sarah your reviewer for this PR 👩‍🚀 

Good job turning on Stickler and bring back the form_tag.

-------

## **Your project is approved! 🎊 🥅 💛**

**You may close this Pull Request and merge the changes**


---------

    
    
**Keep being an awesome dev and good luck with your next project!**

---------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

**Pull request merged**